src/caffe/layers/euclidean_loss_layer.cpp

```diff
@@ -12,12 +12,29 @@ void EuclideanLossLayer<Dtype>::Reshape(
   CHECK_EQ(bottom[0]->count(1), bottom[1]->count(1))
       << "Inputs must have the same dimension.";
   diff_.ReshapeLike(*bottom[0]);
+
+  has_ignore_label_ =
+    this->layer_param_.loss_param().has_ignore_label();
+  if (has_ignore_label_) {
+    ignore_label_ = this->layer_param_.loss_param().ignore_label();
+  }
 }
 
 template <typename Dtype>
 void EuclideanLossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
     const vector<Blob<Dtype>*>& top) {
   int count = bottom[0]->count();
+  if (has_ignore_label_) {
+    Dtype* label_data = bottom[1]->mutable_cpu_data();
+    Dtype* bottom_data = bottom[0]->mutable_cpu_data();
+    for (int i = 0; i < count; ++i) {
+      const int label_value = static_cast<int>(label_data[i]);
+      if (label_value == ignore_label_) {
+        label_data[i] = ignore_label_;
+        bottom_data[i] = ignore_label_;
```

Comment:
Doesn't this violate the contract of layers?  Layers should not be modifying the `bottom` data, since that's the input.